components/script/stylesheet_loader.rs

```diff
@@ -91,6 +96,94 @@ pub struct StylesheetContext {
     resource_timing: ResourceFetchTiming,
 }
 
+impl StylesheetContext {
+    fn unminify_css(&self, data: Vec<u8>, file_url: ServoUrl) -> Vec<u8> {
+        if !self.document.root().window().unminify_css() {
+            return data;
+        }
+
+        let mut style_content = data;
+
+        // Write the minified code to a temporary file and pass its path as an argument
+        // to js-beautify --type css to read from. Meanwhile, redirect the process' stdout into
+        // another temporary file and read that into a string. This avoids some hangs
+        // observed on macOS when using direct input/output pipes with very large
+        // unminified content.
+        let (input, output) = (tempfile::NamedTempFile::new(), tempfile::tempfile());
+        if let (Ok(mut input), Ok(mut output)) = (input, output) {
+            input.write_all(&style_content).unwrap();
+            match Command::new("js-beautify")
+                .arg("--type")
+                .arg("css")
+                .arg(input.path())
+                .stdout(output.try_clone().unwrap())
+                .status()
+            {
+                Ok(status) if status.success() => {
+                    output.seek(std::io::SeekFrom::Start(0)).unwrap();
+                    output.read_to_end(&mut style_content).unwrap();
+                },
+                _ => {
+                    log::warn!(
+                        "Failed to execute js-beautify --type css. Will store unmodified css"
+                    );
+                },
+            }
+        } else {
+            log::warn!("Error creating input and output files for unminify css");
+        }
+
+        let path = match self.document.root().window().unminified_css_dir() {
+            Some(unminified_css_dir) => PathBuf::from(unminified_css_dir),
+            None => {
+                log::warn!("Unminified Css directory not found");
+                return style_content;
+            },
+        };
+
+        let (base, has_name) = match file_url.as_str().ends_with('/') {
+            true => (
+                path.join(&file_url[url::Position::BeforeHost..])
+                    .as_path()
+                    .to_owned(),
+                false,
+            ),
+            false => (
+                path.join(&file_url[url::Position::BeforeHost..])
+                    .parent()
+                    .unwrap()
+                    .to_owned(),
+                true,
+            ),
+        };
+
+        let path = if has_name {
+            // External stylesheet.
+            path.join(&file_url[url::Position::BeforeHost..])
+        } else {
+            // Inline script or url ends with '/'
+            base.join(Uuid::new_v4().to_string())
+        };
+
+        if let Err(e) = create_dir_all(base.clone()) {
```

Comment:
@jdm I was testing this, and on some websites like Google, it fails because the CSS file name is too long, which prevents creating the file. What do you think about shortening or modifying the name to make it work?

Reply:
Very interesting! Let's truncate the name to some reasonable number like 100 characters, I guess?